File: backend/src/modules/users/user.model.ts
import { Schema, model, Document, models } from 'mongoose';
import { Roles, UserRole } from '../../constants/roles';

export interface IUser extends Document {
  firebaseUid: string;
  email: string;
  username: string;
  displayName: string;
  avatarUrl?: string;
  isOnline?: boolean;
  lastSeen?: Date;
  role: UserRole;
  isBanned: boolean;
  createdAt: Date;
  updatedAt: Date;
}

const UserSchema = new Schema<IUser>(
  {
    firebaseUid: { type: String, required: true, unique: true, index: true },
    email: { type: String, required: true, unique: true, lowercase: true, trim: true },
    username: { type: String, required: true, unique: true, lowercase: true, trim: true, index: true },
    displayName: { type: String, required: true, trim: true },
    avatarUrl: { type: String, default: '' },
    isOnline: { type: Boolean, default: false, index: true },
    lastSeen: { type: Date, default: Date.now },
    role: { type: String, enum: Object.values(Roles), default: Roles.USER, index: true },
    isBanned: { type: Boolean, default: false, index: true },
  },
  { timestamps: true }
);

UserSchema.index({ displayName: 'text', username: 'text', email: 'text' });

export const User = models.User || model<IUser>('User', UserSchema);

File: frontend/lib/shared/models/user_model.dart
class UserModel {
  final String id;
  final String firebaseUid;
  final String email;
  final String username;
  final String displayName;
  final String avatarUrl;
  final String statusMessage;
  final bool isOnline;
  final DateTime lastSeen;
  final String role;
  final bool isBanned;

  UserModel({
    required this.id,
    required this.firebaseUid,
    required this.email,
    required this.username,
    required this.displayName,
    required this.avatarUrl,
    required this.statusMessage,
    required this.isOnline,
    required this.lastSeen,
    required this.role,
    required this.isBanned,
  });

  factory UserModel.fromJson(Map<String, dynamic> json) {
    return UserModel(
      id: json['_id'] ?? json['id'] ?? '',
      firebaseUid: json['firebaseUid'] ?? '',
      email: json['email'] ?? '',
      username: json['username'] ?? '',
      displayName: json['displayName'] ?? '',
      avatarUrl: json['avatarUrl'] ?? '',
      statusMessage: json['statusMessage'] ?? 'Hey there! I am using YoumeChat.',
      isOnline: json['isOnline'] ?? false,
      lastSeen: json['lastSeen'] != null ? DateTime.parse(json['lastSeen']) : DateTime.now(),
      role: json['role'] ?? 'User',
      isBanned: json['isBanned'] ?? false,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      '_id': id,
      'firebaseUid': firebaseUid,
      'email': email,
      'username': username,
      'displayName': displayName,
      'avatarUrl': avatarUrl,
      'statusMessage': statusMessage,
      'isOnline': isOnline,
      'lastSeen': lastSeen.toIso8601String(),
      'role': role,
      'isBanned': isBanned,
    };
  }
}
